# 7장. 오류처리

클린 코드 책의 주요 챕터들 중  읽고 싶은 챕터들을 체크해서 7개 정도의 챕터를 읽었는데, 그 중 하나가 7장. 오류처리 였다. 클래스, 객체 관련 챕터를 읽으면서는 클린아키텍처 라는 책을 사고 싶었는데, 잘 참고 아직은 사지 않았다.<br>

Java 개발자는 너무 공부할 게 너무 많다. 가끔은 이런 점이 답답하게 느껴질 때도 많았다. 이제는 그냥 그런가 보다 하고 생각하고 있다. <br>

<br>

## 목차

- [참고자료](#참고자료)
- [Overview](#overview)
- [오류 코드보다 예외를 사용하라](#오류-코드보다-예외를-사용하라)
- [Try, Catch, Finally 문 부터 작성하라](#Try--Catch--Finally-문-부터-작성하라)
- [unchecked-예외를-사용하라](#unchecked-예외를-사용하라)
- [예외에-의미를-제공하라](#예외에-의미를-제공하라)
- [호출자를 고려해 예외 클래스를 정의하라](#호출자를-고려해-예외-클래스를-정의하라)

<br>

## 참고자료

[클린코드 , 로버트 C. 마틴](http://www.yes24.com/Product/Goods/11681152)<br>

<br>

## Overview

7장에서는 깨끗하고 튼튼한 코드에 한걸음 다가가는 단계로 우아하고 고상하게 오류를 처리하는 기법과 고려 사항 몇 가지를 소개한다.<br>

7장에서 이야기하는 주요 골자들을 요약해보면 아래와 같다.<br>

- 오류 코드보다 예외를 사용하라
  - 오류 코드보다는 예외를 사용해야 하는 이유로, caller 측의 코드가 복잡해진다는 단점을 예로 들고 있다.
  - 실제로, 오류코드를 사용하는 예전 방식은 callee 내에서 발생할 수 있는 모든 오류코드들을 caller 내의 if 구문으로 모두 처리해주어야 한다. (caller 와 callee 간의 의존성이 높아진다.)
- Try, Catch, Finally 문 부터 작성하라
- 미확인(unchecked) 예외를 사용하라
- 예외에 의미를 제공하라
- 호출자를 고려해 예외클래스를 정의하라
- 정상 흐름을 정의하라
- null 을 반환하지 마라
- null 을 전달하지 마라

<br>

## 오류 코드보다 예외를 사용하라

- 옛날 언어들에서는 예외를 지원하지 않는 경우가 있었다. 이런 이유로 에러 코드를 직접 정의해 사용하는 경우가 많았다. 책에서는 이렇게 에러 코드를 정의하는 것 보다는 예외를 정의하여 사용하는 것을 권장하고 있다.
- 책에서는 오류 코드보다는 예외를 사용해야 하는 이유로, caller 측의 코드가 복잡해진다는 단점을 예로 들고 있다.
- 실제로, 오류코드를 사용하는 예전 방식은 callee 내에서 발생할 수 있는 모든 오류코드들을 caller 내의 if 구문으로 모두 처리해주어야 한다. (caller 와 callee 간의 의존성이 높아진다.)



**오류코드 또는 오류 플래그를 사용하는 예전 방식** <br>

요즘은 대부분의 프로그래밍 언어에서 예외를 제공한다. 예전에는 예외 대신, 오류 플래그 또는 오류코드를 직접 정의해 사용했었다. 예를 들면 아래의 코드와 같은 방식으로 오류코드를 정의해 사용했었다.<br>

```java
public class DeviceController{
  public void sendShutdown(){
    DeviceHandle handle = getHandle(DEV1);
    
    // 디바이스 상태를 점검 (INVALID 라는 에러 코드를 사용했다.)
    if(handle != Devicehandle.INVALID){
      // 레코드 필드에 디바이스 상태를 저장
      retrieveDeviceRecord(handle);
      // 디바이스가 일시정지 상태가 아니라면 종료한다. (DEVICESUSPENDED) 라는 에러 코드를 사용했다.
      if(record.getStatus() != DEVICESUSPENDED){
        pauseDevice(handle);
        clearDeviceWorkerQueue(handle);
        closeDevice(handle);
      }
      else{
        logger.log("Device suspended. Unable to shut down");
      }
    }
    else{
      logger.log("Invalid handle for " + DEV1.toString());
    }
  }
}
```



이렇게 **오류코드를 사용하는 방식은 caller 측의 코드가 복잡해진다는 단점**이 있다. 또한 **오류코드를 확인하는 것을 caller 측의 코드에 추가하는 것을 잊어버릴 수 있다는 단점** 역시 있다.<br>

<br>

**예외를 사용하도록 로직을 변경하기**<br>

```java
public class DeviceController{
  ...
  
  public void sendShutDown(){
    try{
      tryToShutDown();
    }
    catch(DeviceShutDownError e){
      logger.log(e);
    }
  }
  
  private void tryToShutDown() throws DeviceShutDownError {
    DeviceHandle handle = getHandle(DEV1);
    DeviceRecord record = retrieveDeviceRecord(handle);
    
    pauseDevice(handle);
    clearDeviceWorkerQueue(handle);
    closeDevice(handle);
  }
 
  private DeviceHandle getHandle(DeviceID id){
    // ...
    throw new DeviceShutDownError("Invalid handle for : " + id.toString());
    // ...
  }
  // ...
}
```

<br>

## Try, Catch, Finally 문 부터 작성하라

예외처리시 프로그램 안에 범위를 지정하는 것은 유익한 점이다. try 문에서 예외 발생시 바로 catch 구문으로 넘어가는데, 이런 면세서 try 블록은 트랜잭션과 유사하다고 책에서는 비유하고 있다. 예외가 발생할 수 있는 코드는 try, catch, finally 로 시작하도록 하는 것이 낫다. 이렇게 하면, try 블록에서 무슨 일이 생기든지 caller 로직에서 기대하는 상태를 정의하기 쉬워지게 된다. <br>

<br>

이렇게 try, catch, finally 구문을 개발 버전으로 작성할 때 책에서는 예외가 발생할 수 있는 케이스를 정의해서, 실제 그 예외를 던지는 예외 테스트 케이스를 작성하면서 개발하는 방식을 책에서는 권장하고 있다. (실제로 현업에서도 TDD기반으로 개발을 시작할때 이렇게 시작한다. 물론 테스트 케이스, TC를 늘려갈때는 다른 이야기이긴 하다.) <br>

<br>

실제 개발을 시작할 때, Try, Cactch, Finally 를 사용한 코드를 작성할 때 

- 강제로 예외를 일으키는 테스트 케이스를 작성한 후 
- 테스트 코드 에서는 기대되는 예외가 발생하는지 체크하여 테스트를 통과하도록 코드를 작성하는 방법

이 권장된다.

<br>

ex)

**예제) 테스트코드 (호출자, caller 의 동작을 테스트한다. 기대되는 예외는 StorageException.class.) **<br>

```java
@Test(expected = StorageException.class)
public void retrieveSectionShouldThrowOnInvalidFileName(){
  sectionStore.retrieveSection("invalid - file");
}
```

<br>

**예제) callee 측 코드**<br>

FileNotFoundException 이 발생할 경우 StorageException 을 throw 하고 있다.

```java
public List<RecordedGrip> retrieveSection(String sectionName){
  try{
    FileInputStream stream = new FileInputStream(sectionName);
    stream.close();
  }
  catch(FileNotFoundException e){
    throw new StorageException("retrieval error", e);
  }
  return new ArrayList<RecordedGrip>();
}
```

<br>

이렇게 테스트 케이스 내에서는 예외가 발생할 수 있는 코드를 발생시키고

> sectionStore.retrieveSection("invalid - file");

<br>

이 것이 StorageException.class 예외를 발생하도록 기대하게끔 구문을 추가한다.

> @Test(expected = StorageException.class)

<br>

이 것을 기반으로 retrieveSection 메서드를 정의한다. 위에서 작성한 calller 역할을 하는 테스트 메서드에서는 예외가 발생하게끔하는 인자값인 "invalid - file" 이라는 문자열을 전달해주었다. 따라서 아래의 구문을 거쳐서 StorageExcpetion 을 throw 하게 된다.<br>

> catch(FileNotFoundException e){ <br>
>
> ​	throw new StorageException ("retrival error", e); <br>
>
> } <br>

<br>

## unchecked 예외를 사용하라

> 아주 오래전 자바버전에서는 checked 예외를 사용해왔는데, 이 것은 유연한 프로그램에서도 좋지 않았고 하위 호출단을 수정함으로써, 상위 호출단을 모두 수정해야 한다는 단점이 있었다. 이런 이유로 책에서는 unckeced 예외를 사용할 것을 권장하고 있다.

책에서는 미확인 예외라는 용어로 설명하고 있다. 배우는 입장에서는 IDE에서 바로 표시되는 형식인 unchecked 로 이해하는게 제일 빠르고 편한 방법이라고 생각해서 이번 정리 문서에서는 unchecked 라는 용어만 사용할 예정이다.<br>

<br>

자바의 첫 버전부터 약간은 옛날 까지는 checked 예외(확인된 예외)가 좋은 방식으로 여겨졌다. 이 방식은 메서드 선언시 메서드 내에 메서드가 반환할 예외를 모두 열거하는 방식이다. 또한, 메서드가 반환하는 예외는 메서드 유형 중 일부였다.<br>

<br>

그런데 이 방식은 더는 좋은 방식이 아니다. 다른 언어의 예를 살펴보면  C#, C++, Python, Ruby 모두 checked 예외(확인된 예외)를 제공하지 않는다. checked 예외를 지원하지 않음에도, C#, C++, python, ruby 는 안정적인 소프트웨어를 구현하기에 무리가 없다.<br>

<br>

checked 예외(확인된 예외)는 비용을 치르게 된다. checked 예외(확인된 예외)는 OCP(Open Closed Principle) 을 위반한다. 메서드에서 확인된 예외를 던졌는데, catch 블록이 세 단계 위에 있다면, 그 사이 메서드 모두가 선언부에 해당 예외를 정의해야 한다. 즉, 하위 단계에서 코드를 변경시 상위 메서드 선언부를 모두 고쳐야만 한다.<br>

<br>

## 예외에 의미를 제공하라

- 예외를 던질 때에는 전후 상황을 충분히 덧붙여야 한다. 
- 오류 메시지에 정보를 담아 예외와 함께 던져야 한다. 
- 실패한 연산 이름과 실패 유형도 언급해야 한다.
- 이렇게 하면 catch 블록에서 로깅을 사용해 오류를 기록할 수 있게 된다.

<br>

## 호출자를 고려해 예외 클래스를 정의하라

애플리케이션 내에서 오류를 정의할 때 프로그래머는 오류를 잡아내는 방법에 대해 가장 주의를 기울여야 한다.<br>오류를 분류하는 방법은 수 없이 많은데, 오류가 발생한 위치, 오류가 발생한 컴포넌트, 유형, 디바이스 실패 등으올 분류 가능하다. 이 중에서도 주요한 관심사는 오류를 잡아내는 방법에 주의를 기울여야 한다.<br>

**예) 오류를 나쁘게 분류한 예**

```java
ACMEPort port = ACMEPort(12);

try{
  port.open();
} catch(DeviceResponseException e){
  reportPortError(e);
  logger.log("Device response exception", e);
} catch(ATM1212UnlockedException e){
  reportPortError(e);
  logger.log("Unlock exception", e);
} catch(GMXError e){
  reportPortError(e);
  logger.log("Device response exception");
}
```

<br>

위의 코드는 port 를 open 해서 여러가지 예외 처리 작업들을 단순히 나열해두기만 했다. 단순히 로그를 출력하는 부분과 reportPortError(Exception) 을 호출하는 부분이 단순 반복되고 있다. 그리고, 단순한 코드의 나열보다는 `Port 를 오픈하다가 Exception이 발생했습니다` 하는 것을 명시적으로 나타낸다면 더 좋을 것 같다.<br>

이러한 방식을 감싸기 기법이라고 한다. 감싸기 기법을 통해 Exception 을 처리하는 구문을 표현한 예제는 아래와 같다.<br>

이러한 점을 반영해본 예제는 아래와 같다.<br>

**예) Port open 시의 Exception 을 명확히 구분한 예**

```java
public class LocalPort{
  private ACMEPort innerPort;
  
  public LocalPort(int portNumber){
    innerPort = new ACMEPort(portNumber);
  }
  
  public void open(){
    try{
      innerPort.open();
    }
    catch(DeviceResponseException e){
      throw new PortDeviceFailure(e);
    }
    catch(ATM1212UnlockedException e){
      throw new PortDeviceFailure(e);
    }
    catch(GMXError e){
      throw new PortDeviceFailure(e);
    }
  }
  // ...
}
```

ATM1212UnlockedException 인지, DeviceResponseException인지, GMXError 인지 구분하는 것은 상위호출단에서 모두 결정하기 보다는 Callee 인 LocalPort::open() 메서드에서 결정짓는 것이 코드 간의 의존성을 크게 줄여주기도 하고, 보기에도 좋아보이고, 코드를 조금 더 명확해보이게 해준다.<br>

<br>

이렇게 감싸기 기법을 사용하면 특정 업체가 API를 설계한 방식에 발목잡히지 않는다. 프로그램이 사용하기 편리한 API를 사용하면 그만이기 때문이다.

<br>

## 정상 흐름을 정의하라

여기부터는 내일부터...<br>

새로운 회사로 출근했는데, 집에와서 할것이 많았다. 맥북에 해줘야 하는 것이 너무 많았다...ㄷㄷㄷ <br>

거기다 주제에 퇴근길에 길을 잘못 찾아서 ㅋㅋㅋ 미금역까지 갔다가 다시 청계산 입구로 가지를 않나... 아주 오두방정을 떨면서 퇴근했었다 오늘...<br>

<br>





 











